File: shiftshop-service/src/main/java/com/shiftshop/service/model/services/CatalogService.java
package com.shiftshop.service.model.services;

import com.shiftshop.service.model.common.exceptions.DuplicateInstancePropertyException;
import com.shiftshop.service.model.common.exceptions.InstanceNotFoundException;
import com.shiftshop.service.model.entities.Category;
import com.shiftshop.service.model.entities.Product;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

public interface CatalogService {

    Category addCategory(String name) throws DuplicateInstancePropertyException;

    Category findCategoryById(Long id) throws InstanceNotFoundException;

    List<Category> findAllCategories();

    Category updateCategory(Long id, String name) throws DuplicateInstancePropertyException, InstanceNotFoundException;

    Product addProduct(String name, BigDecimal providerPrice, BigDecimal salePrice, Long categoryId)
            throws DuplicateInstancePropertyException, InstanceNotFoundException;

    Product findProductById(Long id) throws InstanceNotFoundException;

    Block<Product> findProducts(Long categoryId, String keywords, boolean onlyActive,
                                String orderBy, String direction, int page, int size);

    Product updateProduct(Long id, String name, BigDecimal providerPrice, BigDecimal salePrice,
                          String barcode, Long categoryId)
            throws DuplicateInstancePropertyException, InstanceNotFoundException;

    void setActiveProduct(Long id, boolean active) throws InstanceNotFoundException;

    LocalDateTime getLastProductUpdatedTimestamp();

    List<Product> getUpdatedProducts(LocalDateTime lastUpdate);

}

File: shiftshop-spa/src/backend/catalogService.js
import {appFetch, config} from './appFetch';

/* Categories */

export const addCategory = (category, onSuccess, onError, atFinally) =>
    appFetch('/catalog/categories', config('POST', category),
        onSuccess, onError, atFinally);

export const findAllCategories = (onSuccess) =>
    appFetch('/catalog/categories', config('GET'), onSuccess);

export const updateCategory = (id, category, onSuccess, onError, atFinally) =>
    appFetch(`/catalog/categories/${id}`, config('PUT', category),
        onSuccess, onError, atFinally);

/* Products */

export const addProduct = (product, onSuccess, onError, atFinally) =>
    appFetch('/catalog/products', config('POST', product), onSuccess, onError, atFinally);

export const findProducts = ({categoryId, keywords, orderBy, order, onlyActive, page}, onSuccess, atFinally) => {

    let path = `/catalog/products?page=${page}`;

    if (categoryId) {
        path += `&categoryId=${categoryId}`;
    }

    if (keywords && keywords.trim() !== '') {
        path += `&keywords=${keywords.trim()}`;
    }

    if (orderBy && orderBy !== '') {
        path += `&orderBy=${orderBy}`;
    }

    if (order && order !== '') {
        path += `&order=${order}`;
    }

    if (onlyActive) {
        path += `&onlyActive=true`;
    }

    appFetch(path, config('GET'), onSuccess, null, atFinally);
};

export const findProductById = (id, onSuccess, atFinally) =>
    appFetch(`/catalog/products/${id}`, config('GET'), onSuccess, null, atFinally);
